File: C++/convert.cpp
/**
 * @File    : convert.cpp
 * @Brief   : Z 字形变换
 * @Link    : https://leetcode-cn.com/problems/zigzag-conversion/
 * @Author  : Wei Li
 * @Date    : 2021-08-16
*/

#include <iostream>
#include <string>

class Solution
{
public:
    std::string convert(std::string s, int numRows)
    {
        if (numRows == 1)
            return s;

        std::string ret;
        int n = s.size();
        int cycleLen = 2 * numRows - 2;

        for (int i = 0; i < numRows; i++)
        {
            for (int j = 0; j + i < n; j += cycleLen)
            {
                ret += s[j + i];
                if (i != 0 && i != numRows - 1 && j + cycleLen - i < n)
                    ret += s[j + cycleLen - i];
            }
        }
        return ret;
    }
};

// --------------------------------
int main(int argc, char **argv)
{
    std::string str = "PAYPALISHIRING";
    int num_rows = 3;

    Solution solution;
    std::string convert_z = solution.convert(str, num_rows);
    std::cout << "The solution of this problem is " << convert_z << std::endl;

    return 0;
}

File: Python/convert.py
#!/usr/bin/env python3
# encoding: utf-8

"""
@Filename: convert.py
@Function: Z 字形变换
@Link: https://leetcode-cn.com/problems/zigzag-conversion/
@Python Version: 3.8
@Author: Wei Li
@Date：2021-08-16
"""


class Solution:
    def convert(self, s: str, numRows: int) -> str:
        if numRows < 2:
            return s
        res = ["" for _ in range(numRows)]
        i, flag = 0, -1
        for c in s:
            res[i] += c
            if i == 0 or i == numRows - 1:
                flag = -flag
            i += flag
        return "".join(res)


# ----------------------------------
if __name__ == "__main__":
    s = "PAYPALISHIRING"
    numRows = 4

    solution = Solution()
    convert_z = solution.convert(s, numRows)
    print(f"The solution of this problem is {convert_z}")
